File: golden-paths/kotlin/ktor/src/test/kotlin/ContractTest.kt
// Golden-path contract tests — drive the REAL app in-process via Ktor's testApplication.
package golden

import io.ktor.client.request.get
import io.ktor.client.statement.bodyAsText
import io.ktor.http.HttpStatusCode
import io.ktor.server.testing.testApplication
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ContractTest {
    @Test
    fun health_is_ok() = testApplication {
        application { module() }
        val res = client.get("/health")
        assertEquals(HttpStatusCode.OK, res.status)
        assertTrue(res.bodyAsText().contains("ok"))
    }

    @Test
    fun ready_is_ready() = testApplication {
        application { module() }
        val res = client.get("/ready")
        assertEquals(HttpStatusCode.OK, res.status)
        assertTrue(res.bodyAsText().contains("ready"))
    }

    @Test
    fun hello_returns_the_known_payload() = testApplication {
        application { module() }
        val res = client.get("/hello")
        assertEquals(HttpStatusCode.OK, res.status)
        val body = res.bodyAsText()
        assertTrue(body.contains("hello, weyland"))
        assertTrue(body.contains("golden-kotlin-ktor"))
    }

    @Test
    fun metrics_exposes_prometheus() = testApplication {
        application { module() }
        client.get("/hello") // touch the counter so it appears in the exposition
        val res = client.get("/metrics")
        assertEquals(HttpStatusCode.OK, res.status)
        assertTrue(res.bodyAsText().contains("golden_hello_requests"))
    }
}

File: golden-paths/scala/http4s/src/test/scala/ContractTest.scala
// Golden-path contract tests — drive the REAL routes in-process (no server) via http4s.
package golden

import cats.effect.IO
import munit.CatsEffectSuite
import org.http4s.implicits.*
import org.http4s.{Method, Request, Status, Uri}

class ContractTest extends CatsEffectSuite:

  private def get(path: Uri): IO[(Status, String)] =
    Main.routes.run(Request[IO](Method.GET, path)).flatMap(r => r.as[String].map(b => (r.status, b)))

  test("health is ok") {
    get(uri"/health").map { (status, body) =>
      assertEquals(status, Status.Ok)
      assert(body.contains("ok"))
    }
  }

  test("ready is ready") {
    get(uri"/ready").map { (status, body) =>
      assertEquals(status, Status.Ok)
      assert(body.contains("ready"))
    }
  }

  test("hello returns the known payload") {
    get(uri"/hello").map { (status, body) =>
      assertEquals(status, Status.Ok)
      assert(body.contains("hello, weyland"))
      assert(body.contains("golden-scala-http4s"))
    }
  }

  test("metrics exposes prometheus") {
    get(uri"/hello") *> get(uri"/metrics").map { (status, body) =>
      assertEquals(status, Status.Ok)
      assert(body.contains("golden_hello_requests"))
    }
  }
